src/dispmanx_grabber.cpp

```diff
@@ -0,0 +1,194 @@
+#include "dispmanx_grabber.h"
+#include <stdio.h>
+#include <math.h>
+
+#define kprintf if(st->cfg.verbose && st->cfg.logPrintf) st->cfg.logPrintf
+
+static DispManxImageInfo imageInfo[] =
+{
+    { "RGB565", VC_IMAGE_RGB565, 2 },
+    { "RGB888", VC_IMAGE_RGB888, 3 },
+    { "BGR888", VC_IMAGE_BGR888, 3 },
+    { "RGBA16", VC_IMAGE_RGBA16, 2 },
+    { "RGBA32", VC_IMAGE_RGBA32, 4 }
+};
+static size_t imageEntries = sizeof(imageInfo)/sizeof(imageInfo[0]);
+
+void dispmanx_grabber_print_names(void)
+{
+    size_t entry = 0;
+    for (entry = 0; entry < imageEntries; entry++)
+    {
+        printf(" %s", imageInfo[entry].name);
+    }
+}
+
+
+
+int align(int value, int alignLog2)
+{
+    int alignBits = (1 << alignLog2) -1;
+    return (value + alignBits) & ~alignBits;
+}
+
+
+int dispmanx_grabber_init(DispmanxGrabberState *st,
+                          DispmanxGrabberConfig cfg)
+                          
+{
+    int result;
+    memset(st, 0, sizeof(DispmanxGrabberState));
+
+    size_t entry = 0;
+    for (entry = 0; entry < imageEntries; entry++)
+    {
+        if (strcasecmp(cfg.imageTypeName, imageInfo[entry].name) == 0)
+        {
+            st->frameInfo.imageType = imageInfo[entry].type;
+            st->frameInfo.bytesPerPixel =  imageInfo[entry].bytesPerPixel;
+            break;
+        }
+    }
+
+    if (st->frameInfo.imageType == VC_IMAGE_MIN)
+    {
+
+      return DISPMANX_GRABBER_INVALID_IMAGE_NAME;
+    }
+    
+    st->cfg = cfg;
+    
+    bcm_host_init();
+    st->displayHandle = vc_dispmanx_display_open(0);
+    if(!st->displayHandle)
+    {
+        return DISPMANX_GRABBER_DISP_FAIL;
+    }
+    
+    result = vc_dispmanx_display_get_info(st->displayHandle, &st->modeInfo);
+    if (result != 0)
+    {
+        return DISPMANX_GRABBER_MODE_FAIL;
+        
+    }
+    int width = st->modeInfo.width;
+    int height = st->modeInfo.height;
+
+    if (cfg.requestedWidth > 0)
+      {
+        // avoid larger widths than the screen fits
+        cfg.requestedWidth = (cfg.requestedWidth > width ? width : cfg.requestedWidth);
+	
+        if (cfg.requestedHeight == 0)
+	  {
+            double numerator = height * cfg.requestedWidth;
+            double denominator = width;
+            height = (int)ceil(numerator / denominator);
```

Comment:
This bit is super confusing. Presumably what we are trying to do is to work out if we scaled down the width based on the min operation above and, if so, scale the height by the same factor. Maybe it would be clearer to write something along the lines of:

```cpp
double scale = static_cast<double>(cfg.requestedWidth) / static_cast<double>(width);
height = static_cast<int>(ceil(height * scale))
```

(also old C-style casts in a C++ program).

The other thing that makes this confusing is that when read in conjunction with the `if` clause just below, it isn't obvious that because we are in this current `if` statement that we will miss the subsequent one so I was very confused about whether there would be some contamination of the `width`, `height`, `cfg.requestedWidth` and `cfg.requestedHeight` variables. I don't know if there is anything that can be done to improve this but that's how it reads.

Reply:
Yeah, that's all pretty confusing, I'll have a think about it.